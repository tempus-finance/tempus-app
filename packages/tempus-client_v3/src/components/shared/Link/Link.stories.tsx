import { ComponentStory } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import Link from './Link';

export default {
  title: 'Link',
  component: Link,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Template: ComponentStory<typeof Link> = args => (
  <div style={style}>
    <BrowserRouter>
      <Link {...args}>Link</Link>
    </BrowserRouter>
  </div>
);

export const GoogleLink = Template.bind({});
GoogleLink.args = {
  title: 'link to google',
  href: 'https://www.google.com',
};

export const RootLink = Template.bind({});
RootLink.args = {
  title: 'link to root',
  href: '/',
};
