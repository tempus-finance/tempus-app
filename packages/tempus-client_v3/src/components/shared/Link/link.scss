.tc__link {
  text-decoration: none;
}
