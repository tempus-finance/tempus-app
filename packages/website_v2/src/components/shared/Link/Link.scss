.tw__link {
  text-decoration: none;
  color: inherit;
}
