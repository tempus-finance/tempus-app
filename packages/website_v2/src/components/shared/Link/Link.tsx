import { FC, forwardRef, memo, HTMLProps, useMemo } from 'react';
import { Link as InternalLink } from 'react-router-dom';

import './Link.scss';

interface LinkProps {
  className?: string;
  title?: string;
  href?: string;
  onClick?: () => void;
}

const Link = forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const { children, className = '', title, href = '', onClick } = props;

  const isExternal = useMemo(() => href.includes('://') && new URL(href).origin !== window.location.origin, [href]);

  const isHash = useMemo(() => href.startsWith('#'), [href]);

  return (
    <>
      {isExternal && (
        <a
          ref={ref}
          rel="external noreferrer nofollow"
          target="_blank"
          className={`tw__link ${className}`}
          title={title}
          href={href}
          onClick={onClick}
        >
          {children}
        </a>
      )}
      {isHash && (
        <a ref={ref} className={`tw__link ${className}`} title={title} href={href} onClick={onClick}>
          {children}
        </a>
      )}
      {!isExternal && !isHash && (
        <InternalLink ref={ref} className={`tw__link ${className}`} title={title} to={href} onClick={onClick}>
          {children}
        </InternalLink>
      )}
    </>
  );
});

export default memo(Link as FC<LinkProps & HTMLProps<HTMLAnchorElement>>);
