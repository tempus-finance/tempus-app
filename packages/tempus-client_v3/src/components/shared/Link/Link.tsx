import React, { FC, forwardRef, memo, HTMLProps, useMemo } from 'react';
import { Link as InternalLink } from 'react-router-dom';

import './link.scss';

interface LinkProps {
  className?: string;
  title?: string;
  href?: string;
  onClick?: () => void;
}

const Link = forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const { children, className = '', title, href = '', onClick } = props;

  const isExternal = useMemo(() => href.includes('://') && new URL(href).origin !== window.location.origin, [href]);

  return isExternal ? (
    <a
      ref={ref}
      rel="external noreferrer nofollow"
      target="_blank"
      className={`tc__link ${className}`}
      title={title}
      href={href}
      onClick={onClick}
    >
      {children}
    </a>
  ) : (
    <InternalLink ref={ref} className={`tc__link ${className}`} title={title} to={href} onClick={onClick}>
      {children}
    </InternalLink>
  );
});

export default memo(Link as FC<LinkProps & HTMLProps<HTMLAnchorElement>>);
