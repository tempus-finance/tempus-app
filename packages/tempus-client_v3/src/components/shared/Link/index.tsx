import React, { ForwardedRef, forwardRef, memo, ReactNode, useMemo } from 'react';
import { Link as InternalLink } from 'react-router-dom';

import './link.scss';

export interface LinkProps {
  children: ReactNode;
  className?: string;
  title?: string;
  href?: string;
  onClick?: () => void;
}

const Link = forwardRef((props: LinkProps, ref: ForwardedRef<HTMLAnchorElement>) => {
  const { children, className = '', title, href = '', onClick } = props;

  const isExternal = useMemo(() => href.includes('://') && new URL(href).origin !== window.location.origin, [href]);

  return isExternal ? (
    <a
      ref={ref}
      rel="external noreferrer nofollow"
      target="_blank"
      className={`tc__link ${className}`}
      title={title}
      href={href}
      onClick={onClick}
    >
      {children}
    </a>
  ) : (
    <InternalLink ref={ref} className={`tc__link ${className}`} title={title} to={href} onClick={onClick}>
      {children}
    </InternalLink>
  );
});

export default memo(Link);
