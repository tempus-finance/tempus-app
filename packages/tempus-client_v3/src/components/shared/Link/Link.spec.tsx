import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Link, { LinkProps } from './Link';

const mockOnClick = jest.fn();

const defaultProps: LinkProps = {
  onClick: mockOnClick,
};

const subject = (props: LinkProps) =>
  render(
    <BrowserRouter>
      <Link {...props}>Link</Link>
    </BrowserRouter>,
  );

describe('Link', () => {
  it('renders a link to google', () => {
    const props = {
      ...defaultProps,
      className: 'tc__external-link',
      title: 'link to google',
      href: 'https://www.google.com',
    };
    const { getByRole } = subject(props);

    const link = getByRole('link');

    expect(link).not.toBeNull();
    expect(link).toHaveAttribute('title', props.title);
    expect(link).toHaveAttribute('href', props.href);
    expect(link.getAttribute('class')).toContain(props.className);

    expect(link).toMatchSnapshot();
  });

  it('renders a link to internal site', () => {
    const props = {
      ...defaultProps,
      className: 'tc__internal-link',
      title: 'link to team page',
      href: '/team',
    };
    const { getByRole } = subject(props);

    const link = getByRole('link');

    expect(link).not.toBeNull();
    expect(link).toHaveAttribute('title', props.title);
    expect(link).toHaveAttribute('href', props.href);
    expect(link.getAttribute('class')).toContain(props.className);

    expect(link).toMatchSnapshot();
  });

  it('renders an empty link', () => {
    const { getByRole } = subject(defaultProps);

    const link = getByRole('link');

    expect(link).not.toBeNull();
    expect(link).not.toHaveAttribute('title');
    expect(link).toHaveAttribute('href', '/');

    expect(link).toMatchSnapshot();
  });
});
