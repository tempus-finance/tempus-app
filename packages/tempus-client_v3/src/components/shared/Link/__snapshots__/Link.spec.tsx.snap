// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders a link to google 1`] = `
<a
  class="tc__link tc__external-link"
  href="https://www.google.com"
  rel="external noreferrer nofollow"
  target="_blank"
  title="link to google"
>
  Link
</a>
`;

exports[`Link renders a link to internal site 1`] = `
<a
  class="tc__link tc__internal-link"
  href="/team"
  title="link to team page"
>
  Link
</a>
`;

exports[`Link renders an empty link 1`] = `
<a
  class="tc__link "
  href="/"
>
  Link
</a>
`;
